inference-engine/src/hetero_plugin/hetero_executable_network.cpp

```diff
@@ -312,14 +347,377 @@ HeteroExecutableNetwork::HeteroExecutableNetwork(const InferenceEngine::ICNNNetw
         auto deviceName = d._device;
         auto metaDevices = _heteroPlugin->GetDevicePlugins(deviceName, cfg);
         assert(metaDevices.size() == 1);
-
         auto loadConfig = metaDevices[deviceName];
         d._network = _heteroPlugin->GetCore()->LoadNetwork(d._clonedNetwork, deviceName, loadConfig);
     }
 
     networks = std::move(descs);
 }
 
+template<typename T>
+using NodeMap = std::unordered_map<ngraph::Node*, T>;
+
+void HeteroExecutableNetwork::InitNgraph(const InferenceEngine::ICNNNetwork& network_) {
+    auto function = network_.getFunction();
+    auto networkPtr = cloneNetwork(network_);
+    auto& network = *networkPtr;
+    auto itDumpDotFile = _config.find(HETERO_CONFIG_KEY(DUMP_GRAPH_DOT));
+    bool dumpDotFile = itDumpDotFile != _config.end() ? (itDumpDotFile->second == YES) : false;
+#ifndef NDEBUG
+    dumpDotFile  = true;
+#endif
+
+    QueryNetworkResult queryNetworkResult;
+    auto orderedOps = function->get_ordered_ops();
+    orderedOps.erase(
+        std::remove_if(std::begin(orderedOps), std::end(orderedOps), [] (const std::shared_ptr<ngraph::Node>& node) {
+            return node->is_constant();
+        }),
+        std::end(orderedOps));
+    bool allEmpty = true;
+    for (auto&& node : orderedOps) {
+        auto& nodeInfo = node->get_rt_info();
+        auto itInfo = nodeInfo.find("affinity");
+        if (itInfo != nodeInfo.end()) {
+            queryNetworkResult.supportedLayersMap.emplace(
+                node->get_friendly_name(),
+                Parameter{itInfo->second}.as<std::string>());
+            allEmpty = false;
+        }
+    }
+
+    if (queryNetworkResult.supportedLayersMap.empty()) {
+        auto it = _config.find("TARGET_FALLBACK");
+        if (it != _config.end()) {
+            _heteroPlugin->QueryNetwork(network_, _config, queryNetworkResult);
+        } else {
+            THROW_IE_EXCEPTION << "The 'TARGET_FALLBACK' option was not defined for heterogeneous plugin";
+        }
+    }
+
+    using Input = ngraph::Input<ngraph::Node>;
+    using NodeSet = std::unordered_set<ngraph::Node*>;
+    using InputSet = std::set<Input>;
+
+    auto InputNode  = [] (const ngraph::Input<ngraph::Node>& input) {
+        return input.get_source_output().get_node();
+    };
+
+    auto NoConstants = [] (std::vector<ngraph::Input<ngraph::Node>>&& inputs) {
+        std::vector<ngraph::Input<ngraph::Node>> result;
+        for (auto&& input : inputs) {
+            if (!(input.get_source_output().get_node()->is_constant())) {
+                result.emplace_back(std::move(input));
+            }
+        }
+        return result;
+    };
+
+    // Set parameters affinity
+    for (auto&& node : function->get_parameters()) {
+        if (!contains(queryNetworkResult.supportedLayersMap, node->get_friendly_name())) {
+            auto& outputNodeName = node->output(0).get_target_inputs().begin()->get_node()->get_friendly_name();
```

Comment:
Why you check only one consumer? Parameter can have multiple consumers.

Reply:
There is no selection cretaria for such case. So the first consumers affinity is good enogh here.